enhancements/update/distribute-secondary-metadata-as-container-image.md

```diff
@@ -0,0 +1,197 @@
+
+### Continuous build and release
+Can CPaaS handle continuous build and release from master?
+
+## Summary
+The [cincinnati-graph-data GitHub repository][cincinnati-graph-data] will be continuously packaged and released in form of a container image by the means of integration with CPaaS.
+This image will be available for public consumption, and will be used by the publicly available Cincinnati instances operated by Red Hat.
+The image can be transferred and served to Cincinnati instances operated by any party, including but not limited to network-restricted environments.
+The source of the container image, namely the cincinnati-graph-data, will remain publicly available and its workflows unaffected by this enhancement.
+
+## Motivation
+
+### Goals
+1. The container image will become the primary way of consuming and distributing the content of the cincinnati-graph-data repository.
+2. As long as the primary release metadata will be packaged, distributed and consumed in form of container images as well, the transport mechanism required for data consumption by Cincinnati, as well as mirroring and serving its consumed content, will be solely [Docker Registry HTTP API V2][docker-registry-http-api-v2] protocol.
+
+
+### Non-Goals
+1. Changes to the GitHub repository workflow or data format.
+2. Implementation details about the streamlining of the mirroring process for primary and secondary release metadata.
+
+## Proposal
+
+### User Stories
+
+#### Content mirroring for Cincinnati deployments
+A user might want to mirror all content which is required by Cincinnati.
+
+1. The user sets up a container image registry in the target environment.
+2. Optionally, if the target environment does not have internet access, the user downloads the release container images, the desired graph-data container image, and the images' verification signatures to a transfer medium.
+3. The user pushes the release, graph-data, and signature images to the target container image registry
+4. The user can now configure the Cincinnati instance in the target environment to consume the content from the available container image registry
+
+### Implementation Details/Notes/Constraints
+
+#### CPaaS Build configuration
+1. Every change to the master branch of the cincinnati-graph-data results in a new container image containing the unmodified working tree files of the new revision.
+  This will allow the plugin to continuously grab the latest version without being informed that a latest version exists.
+
+2. The image will be signed by an official Red Hat key, and a container image including the signature for the latest image will be produced.
+
+3. The image will be pushed to a publicly available container image registry under a well-known repository name, and it's `latest` tag will be updated to the latest revision.
+
+4. The signature verification image will be pushed to the same repository similarly, under the tag `latest-signature`.
```

Comment:
This is horrifying.  But if your constraint is "everything we push has to land in the registry", it may be your only option.  Are you trying to push it with some particular media type?  Or are you dropping it into a layer in a shell container image (if so, you probably have to define a well-known path inside the layer filesystem)?

Reply:
Right, I had the race issue in my head at one point but forgot about it when writing the enhancement, thanks for bringing it up.

I recognize the race condition exists, and I can see that it's practically possible be experienced. I think it's worth exploring options to solve it.

(a) Always update the `latest-signature` tag before the `latest` tag, maintain a `previous-signature` tag, and have Cincinnati try to verify with the latest and previous signature.
(b) Create a third image which merely references to the data and signature images. E.g. without relying on specific container metadata, this third container image could contain two files named `image` and `signatures`, which contain the digests of these images.

Reply:
Maybe just put a label on the first image that contains the tag you'll use for the signature image?

Reply:
What type of labels are you referring to?

Reply:
https://docs.docker.com/engine/reference/builder/#label

As seen for example if you run `skopeo inspect docker://quay.io/fedora/fedora:31-x86_64`

Reply:
Ah these, thanks! I think that'll work, as the information is updated atomically.